core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1124,19 +1120,20 @@ private void sendResponse(Version nodeVersion, Channel channel, final TransportR
             final TransportResponseOptions finalOptions = options;
             Runnable onRequestSent = () -> { // this might be called in a different thread
                 try {
-                    toRelease.close();
+                    bStream.close();
                 } finally {
                     transportServiceAdapter.onResponseSent(requestId, action, response, finalOptions);
                 }
             };
             addedReleaseListener = internalSendMessage(channel, reference, onRequestSent);
         } finally {
             try {
-                IOUtils.close(stream);
```

Comment:
The close logic is confusing. I pushed another commit that implements what you've suggested.

In the new commit, `buildMessage` will call close on the stream. Regardless of the stream type, the base stream (ReleasableBytesStreamOutput/BytesStreamOutput) will only flushed while the other levels will be closed so that compression works.